### PR TITLE
Support for Snappy PR# 2420

### LINKED ISSUE #2420: Snappy Compression support
**Is your feature request related to a problem? Please describe.**

Pipeline author wants to read Read Parquet files on AWS S3 with Snappy Compression.  Currently S3 (via abstract files source) only supports gzip

**Describe the solution you'd like**

Build in support for other compression such as:

- Snappy

### PR DESCRIPTION
**Pull request by ashoktelukuntla**
### Description
Added Snappy compression support 
 
### Issues Resolved
resolves #2420
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by asuresh8 — DISMISSED** [review]

File: data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/configuration/CompressionOption.java [hunk 1]
```diff
@@ -18,6 +19,7 @@
 public enum CompressionOption {
     NONE("none", new NoneCompressionEngine()),
     GZIP("gzip", new GZipCompressionEngine()),
+    SNAPPY("snappy", new SnappyCompressionEngine()),
```
Comment (dinujoh) [inline data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/configuration/CompressionOption.java:22]: Do we need to update existing test ?

Comment (ashoktelukuntla) [reply data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/configuration/CompressionOption.java:22]: Updated existing test and committed

File: data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/compression/AutomaticCompressionEngine.java [hunk 2]
```diff
@@ -21,7 +21,9 @@ public InputStream createInputStream(final String s3Key, final InputStream respo
     private CompressionOption getCompressionOption(final String s3Key) {
         if (s3Key.endsWith(".gz")) {
             return CompressionOption.GZIP;
-        } else {
+        } else if(s3Key.endsWith(".snappy.parquet") || s3Key.endsWith(".snappy")){
```
[the remark this answers is not in the document]
Comment (ashoktelukuntla) [reply data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/compression/AutomaticCompressionEngine.java:24]: Added unit test for both the options

File: data-prepper-plugins/s3-source/src/test/java/org/opensearch/dataprepper/plugins/source/compression/AutomaticCompressionEngineTest.java [hunk 3]
```diff
@@ -71,4 +73,33 @@ void createInputStream_with_automatic_and_compressed_should_return_instance_of_G
...
+    @Test
+    void createInputStream_with_automatic_and_compressed_should_return_instance_of_SnappyInputStream() throws IOException {
+         snappyTest(".snappy");
+    }
+
+    @Test
+    void createInputStream_with_automatic_and_compressed_should_return_instance_of_SnappyInputStreamParquet() throws IOException {
+        snappyTest(".snappy.parquet");
+    }
+    public void snappyTest(String fileExtension) throws IOException {
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/test/java/org/opensearch/dataprepper/plugins/source/compression/AutomaticCompressionEngineTest.java:86]: You can make this a parameterized test.

```
@ParameterizedTest
@ValueSource(strings = { ".snappy", ".snappy.parquet"}
void createInputStream_with_automatic_and_compressed_should_return_instance_of_SnappyInputStream(final String fileExtension) {
  // The code from inside `snappyTest`.
}
```

File: data-prepper-plugins/s3-source/src/test/java/org/opensearch/dataprepper/plugins/source/configuration/SnappyCompressionOptionTest.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+class SnappyCompressionOptionTest {
+    @ParameterizedTest
+    @EnumSource(CompressionOption.class)
+    @CsvSource({"SNAPPY"})
+    void fromOptionValue(final CompressionOption option) {
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/test/java/org/opensearch/dataprepper/plugins/source/configuration/SnappyCompressionOptionTest.java:21]: I think you can remove this test entirely. It is already covered in `CompressionOptionTest`.

https://github.com/opensearch-project/data-prepper/blob/5110699c993d4e17a56ae7d1f8c11def0429c0d9/data-prepper-plugins/s3-source/src/test/java/org/opensearch/dataprepper/plugins/source/configuration/CompressionOptionTest.java#L17-L22

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Comment by dinujoh** [comment]
Thanks the changes look good. We need to update the [README](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/s3-source/README.md) for Snappy compression support.

File: data-prepper-plugins/s3-source/src/test/java/org/opensearch/dataprepper/plugins/source/compression/AutomaticCompressionEngineTest.java (hunk 3 above)
Comment (ashoktelukuntla) [reply data-prepper-plugins/s3-source/src/test/java/org/opensearch/dataprepper/plugins/source/compression/AutomaticCompressionEngineTest.java]: Addressed and now Junit has parameterized test

File: data-prepper-plugins/s3-source/src/test/java/org/opensearch/dataprepper/plugins/source/configuration/SnappyCompressionOptionTest.java (hunk 4 above)
Comment (ashoktelukuntla) [reply data-prepper-plugins/s3-source/src/test/java/org/opensearch/dataprepper/plugins/source/configuration/SnappyCompressionOptionTest.java]: Removed it

**Review by dinujoh — APPROVED** [review]

**Comment by dlvenable** [comment]
> The changes look good! We need to update the [README](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/s3-source/README.md) for Snappy compression support.

This is true. Also, @ashoktelukuntla , please create a new issue in the https://github.com/opensearch-project/documentation-website repository to track this new feature.

File: data-prepper-plugins/s3-source/README.md [hunk 5]
```diff
@@ -22,6 +22,8 @@ The `compression` property defines how to handle compressed S3 objects. It has t
 
 * `none` - The file is not compressed.
 * `gzip` - Apply GZip de-compression on the S3 object.
+* `.snappy` - Apply Snappy de-compression on the S3 object.
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/README.md:25]: There is only one new option - `snappy`. I think combine these lines into:

```
* `snappy` - Apply Snappy decompression on the S3 object.
```

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/s3-source/README.md (hunk 5 above)
Comment (ashoktelukuntla) [reply data-prepper-plugins/s3-source/README.md:25]: Thanks.. updated to only one option

**Review by dinujoh — DISMISSED** [review]

File: data-prepper-plugins/s3-source/README.md (hunk 5 above)
Comment (dinujoh) [reply data-prepper-plugins/s3-source/README.md:25]: Looks like we have '.' as prefix. This should just be `snappy`

**Review by dinujoh — DISMISSED** [review]

File: data-prepper-plugins/s3-source/README.md (hunk 5 above)
Comment (ashoktelukuntla) [reply data-prepper-plugins/s3-source/README.md]: Thanks for the catch- fixed it

**Review by dlvenable — APPROVED** [review]

**Comment by dlvenable** [comment]
Once the DCO is fixed, we can merge.

**Review by dlvenable — APPROVED** [review]

**Review by dinujoh — APPROVED** [review]

**Pull request merged**